core/primitives/src/time.rs

```diff
@@ -0,0 +1,171 @@
+use std::default::Default;
+use std::sync::{Arc, Mutex};
+use std::time;
+
+use arc_swap::ArcSwap;
+use chrono;
+use once_cell::sync::Lazy;
+
+pub use chrono::Utc;
+pub use std::time::Instant;
+
+use std::collections::{HashMap, VecDeque};
+use std::thread::ThreadId;
+
+pub struct MockTimeSingletonPerThread {
+    utc: VecDeque<chrono::DateTime<chrono::Utc>>,
+    durations: VecDeque<time::Duration>,
+    utc_call_count: u64,
+    instant_call_count: u64,
+    instant: time::Instant,
+}
+
+pub struct MockTimeSingleton {
+    threads: HashMap<ThreadId, MockTimeSingletonPerThread>,
+}
+
+impl MockTimeSingletonPerThread {
+    pub fn reset(&mut self) {
+        self.utc.clear();
+        self.durations.clear();
+        self.utc_call_count = 0;
+        self.instant_call_count = 0;
+        self.instant = time::Instant::now();
+    }
+}
+
+impl Default for MockTimeSingletonPerThread {
+    fn default() -> Self {
+        Self {
+            utc: VecDeque::with_capacity(16),
+            durations: VecDeque::with_capacity(16),
+            utc_call_count: 0,
+            instant_call_count: 0,
+            instant: time::Instant::now(),
+        }
+    }
+}
+
+impl Default for MockTimeSingleton {
+    fn default() -> Self {
+        Self { threads: HashMap::default() }
+    }
+}
+
+static SINGLETON: Lazy<ArcSwap<Mutex<MockTimeSingleton>>> =
+    Lazy::new(|| ArcSwap::from_pointee(Mutex::new(MockTimeSingleton::new())));
+
+impl MockTimeSingleton {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn get() -> Arc<Mutex<MockTimeSingleton>> {
+        SINGLETON.load_full()
+    }
+
+    pub fn set(value: MockTimeSingleton) {
+        SINGLETON.store(Arc::new(Mutex::new(value)))
+    }
+
+    pub fn add_utc(&mut self, mock_date: chrono::DateTime<chrono::Utc>) {
+        self.current_mut().unwrap().utc.push_back(mock_date);
+    }
+
+    pub fn pop_utc(&mut self) -> Option<chrono::DateTime<chrono::Utc>> {
+        let instance = self.current_mut().unwrap();
+        instance.utc_call_count += 1;
+        instance.utc.pop_front()
+    }
+
+    pub fn pop_instant(&mut self) -> Option<time::Instant> {
+        let instance = self.current_mut()?;
+        instance.instant_call_count += 1;
+        let x = instance.durations.pop_front();
+        match x {
+            Some(t) => instance.instant.checked_add(t),
+            None => None,
+        }
+    }
+
+    pub fn current_mut(&mut self) -> Option<&mut MockTimeSingletonPerThread> {
+        let handle = std::thread::current();
+        let id = handle.id();
+        if !self.threads.contains_key(&id) {
+            self.threads.insert(id, MockTimeSingletonPerThread::default());
+        }
+        self.threads.get_mut(&id)
+    }
+
+    pub fn current(&self) -> Option<&MockTimeSingletonPerThread> {
```

Comment:
Feels like significant amount of stuff in this file doesn't need to be `pub`. Right now, it's pretty hard to see what is the public interface I should use from other crates, and what is a private impl detail.